aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -717,10 +727,11 @@ module aptos_framework::delegation_pool {
         };
         coin::transfer<AptosCoin>(stake_pool_owner, delegator_address, amount);
 
-        // commit withdrawal of possibly inactive stake to the `total_coins_inactive`
+        // commit withdrawal of inactive stake to the `total_coins_inactive`
         // known by the delegation pool in order to not mistake it for slashing at next synchronization
-        let (_, inactive, _, _) = stake::get_stake(pool_address);
-        pool.total_coins_inactive = inactive;
+        if (withdrawing_inactive_stake) {
```

Comment:
Can you explain this more?

Reply:
We would like to be more explicit when updating total inactive coins known by the DelegationPool.
When withdrawing pending_inactive stake `total_coins_inactive` should be left untouched and this would be cleaner code.
Also there is the following case: lockup cycle ended on StakePool and inactivated all pending_inactive stake (which would be detected at the next synchronization). As `calculate_stake_pool_drift` is called before `buy_in_pending_inactive_shares` (which in turn calls `withdraw_internal`) at `synchronize_delegation_pool`, then the lockup cycle is still successfully detected. But the invariant that `a new lockup is always detected` is scattered across multiple functions.

Reply:
Let me double check if I understand this correctly:
In the existing implementation, after calling `withdraw_internal`, actually `DelegationPool` is in a wrong state - `pool.total_coins_inactive` is synced but `pool.observed_lockup_cycle.index` is still the last OLC.
Let me give an example:
Assume `reward_rate = 0`.
1. In the beginning, `DelegationPool = {total_coins_inactive: 200, OLC: x} StakePool={pending_inactive: 0, inactive:200}`
2. Delegator A unlocks 100 at OLC x. An entry is created in `inactive_shares`. `DelegationPool = {total_coins_inactive: 200, OLC: x} StakePool={pending_inactive: 100, inactive:200}`
3. A lockup period is passed. `DelegationPool = {total_coins_inactive: 200, OLC: x} StakePool={pending_inactive: 0, inactive: 300}`
4. `withdraw_internal(DelegationPool, A, MAX_U64)` is called. `DelegationPool = {total_coins_inactive: 200, OLC: x} StakePool={pending_inactive: 0, inactive: 200}`. 100 is transferred to A.

In this state, we cannot detect if a cycle has passed. So calling `withdraw_internal` needs to be very careful - you have to detect if there is a new cycle before calling it. Currently everywhere follows this pattern but this is very likely to cause bugs.

After this PR, 4. will become
`withdraw_internal(DelegationPool, A, MAX_U64)` is called. `DelegationPool = {total_coins_inactive: 100, OLC: x} StakePool={pending_inactive: 0, inactive: 200}`. 100 is transferred to A.
This is a valid state. Calling `withdraw_internal` is safe.

However, this PR **cannot** cover all the cases because there is **no guarantee** that `pool.total_coins_inactive >= amount`. `pool.total_coins_inactive - amount` might be negative and cause an abort. For example:
1. In the beginning, `DelegationPool = {total_coins_inactive: 0, OLC: x} StakePool={pending_inactive: 0, inactive:0}`
2. Delegator A unlocks 100 at OLC x. An entry is created in `inactive_shares`. `DelegationPool = {total_coins_inactive: 0, OLC: x} StakePool={pending_inactive: 100, inactive:0}`
3. A lockup period is passed. `DelegationPool = {total_coins_inactive: 0, OLC: x} StakePool={pending_inactive: 0, inactive: 100}`
4. withdraw_internal(DelegationPool, A, MAX_U64)` is called. `pool.total_coins_inactive - amount = -100`. Abort.

I would suggest to also update OLC inside `withdraw_internal` instead of `pool.total_coins_inactive - amount`. The state change actually happens at OLC x+1 so it doesn't make sense to update `pool.total_coins_inactive` but not update OLC.